mail_base/static/src/js/base.js

```diff
@@ -0,0 +1,1083 @@
+                }
+            });
+        } else {
+            new Model(res_model).call('get_formview_id', [res_id, session.context]).then(function (view_id) {
+                redirect_to_document(res_model, res_id, view_id);
+            });
+        }
+    };
+chat_manager.get_channels_preview = function (channels) {
+        var channels_preview = _.map(channels, function (channel) {
+            var info = _.pick(channel, 'id', 'is_chat', 'name', 'status', 'unread_counter');
+            info.last_message = _.last(channel.cache['[]'].messages);
+            if (!info.is_chat) {
+                info.image_src = '/web/image/mail.channel/'+channel.id+'/image_small';
+            } else if (channel.direct_partner_id) {
+                info.image_src = '/web/image/res.partner/'+channel.direct_partner_id+'/image_small';
+            } else {
+                info.image_src = '/mail/static/src/img/smiley/avatar.jpg';
+            }
+            return info;
+        });
+        var missing_channels = _.where(channels_preview, {last_message: undefined});
+        if (!channels_preview_def) {
+            if (missing_channels.length) {
+                var missing_channel_ids = _.pluck(missing_channels, 'id');
+                channels_preview_def = ChannelModel.call('channel_fetch_preview', [missing_channel_ids], {}, {shadow: true});
+            } else {
+                channels_preview_def = $.when();
+            }
+        }
+        return channels_preview_def.then(function (channels) {
+            _.each(missing_channels, function (channel_preview) {
+                var channel = _.findWhere(channels, {id: channel_preview.id});
+                if (channel) {
+                    channel_preview.last_message = cls.add_message(channel.last_message);
+                }
+            });
+            return _.filter(channels_preview, function (channel) {
+                return channel.last_message;  // remove empty channels
+            });
+        });
+    };
+chat_manager.get_message_body_preview = function (message_body) {
+        return cls.parse_and_transform(message_body, inline);
+    };
+chat_manager.search_partner = function (search_val, limit) {
+        return PartnerModel.call('im_search', [search_val, limit || 20], {}, {shadow: true}).then(function(result) {
+            var values = [];
+            _.each(result, function(user) {
+                var escaped_name = _.escape(user.name);
+                values.push(_.extend(user, {
+                    'value': escaped_name,
+                    'label': escaped_name
+                }));
+            });
+            return values;
+        });
+    };
+chat_manager.send_native_notification = cls.send_native_notification;
+chat_manager.bus.on('client_action_open', null, function (open) {
+    client_action_open = open;
+});
+
+chat_manager.is_ready = cls.start();
+return {
+    chat_manager: chat_manager,
+    MailTools: MailTools
```

Comment:
make MailTools similar to chat_manager, i.e. just object. Because otherwirse you cannot extend it